### ISSUE TITLE
Filtering and sorting on inherited properties cannot be translated to SQL with EF Core

### ISSUE THREAD

**Issue by SuperGouge** [Biarity/Sieve#64]
Properties declared on a base class cannot be used in filters and sorts when applying them to a query of a derived type. The expressions generated by Sieve are unable to be translated to SQL by EF Core (and probably some other providers using the same LINQ parser). This causes a client evaluation of the query and depending on the configuration and version, this will either cause huge performance issues or simply throw an exception.

Here is an example illustrating exactly what the problem is:

```
public class Foo
{
    public int Id { get; set; }
    public string Bar { get; set; }
}

public class MyDbContext : DbContext
{
    public DbSet<Foo> Foos { get; set; }
}
```
```
public class FooModel
{
    public string Bar { get; set; }
}

public class FooBarModel : FooModel
{
    public string BarFoo { get; set; }
}
```
```
public IList<FooBarModel> GetModels(MyDbContext dbContext, ISieveProcessor sieveProcessor, SieveModel sieveModel)
{
    var query = dbContext.Foos
        .Select(f => new FooBarModel
        {
            Bar = f.Bar,
            BarFoo = f.Bar + f.Bar
        });
    query = sieveProcessor.Apply(sieveModel, query, applyPagination: false);
    return query.ToList();
}
```

```
var dbContext = GetMyDbContext();
var sieveProcessor = GetSieveProcessor();

var validSieveModel = new SieveModel
{
    Filters = "BarFoo==test"
};
var validResult = GetModels(dbContext, sieveProcessor, validSieveModel); // Will be correctly translated to SQL.

var invalidSieveModel = new SieveModel
{
    Filters = "Bar==test"
};
var invalidResult = GetModels(dbContext, sieveProcessor, invalidSieveModel); // Cannot be translated to SQL. Will generate a Microsoft.EntityFrameworkCore.Query.QueryClientEvaluationWarning or throw an exception depending on version/config.
```

More information as well as a possible solution can be found [here](https://github.com/aspnet/EntityFrameworkCore/issues/12230) and [here](https://www.re-motion.org/jira/browse/RMLNQ-121). It looks like the expressions manually generated by Sieve would need to be adapted to reference the properties with the right type.

**Comment by zackattackz** [comment]
Has this feature been implemented yet, or if not then has it been in any forks of Sieve that you know of?

**Comment by hasanmanzak** [comment]
Can we bump this up? I really didn't understand what was problem here, sorry.

```
var validSieveModel = new SieveModel
{
    Filters = "BarFoo==test"
};
```

works but

```
var invalidSieveModel = new SieveModel
{
    Filters = "Bar==test"
};
```

does not?! Honestly, I was expecting the other way around :) There's a known issue with the interfaces that I know of, but it is fixed and waiting to be reviewed. I never encountered such issue with concrete types. Can someone provide more info on this?

**Comment by zackattackz** [comment]
I'm not too familiar with Sieve exactly, I've just been scouting different filtering libraries to find one that best fits my needs - so forgive my lack of knowledge here.

I'm essentially looking for the ability to have filters for a type hierarchy (which is [supported](https://learn.microsoft.com/en-us/ef/core/modeling/inheritance#entity-type-hierarchy-mapping) by ef core)

The filters should be able to be applied to check against specific instances of a sub type in the hierarchy, and many could be applied across the different sub types possible.

If this isn't too clear I could try to explain further with examples, thank you for your time.

**Comment by hasanmanzak** [comment]
Sieve is just an `Expression Tree` generator. Actual `Linq-To-Sql` operations is conducted by `Query Providers`. If provider of an `IQueryable` is a `Linq-To-Sql Provider` then the expression tree is converted to corresponding sql statements by that provider. So it is in EF's responsibility to generate correct sql statements, in your case.

I use sieve in the production with both Linq-To-Sql (with NHibernate) & Linq-To-Entity (In memory). and with different type hierarchies.

As I mentioned before, there is one particular downside Sieve has, it can't handle interface hierarchies correctly which I fixed recently.

I strongly encourage you to give Sieve a go ;) I think there's so much room to grow.

**Comment by zackattackz** [comment]
Thank you for the clarification but I believe maybe I'm not describing my needs the best way. I would like for the Expression Tree to be generated with filters of sub classes in mind. (Using [Expression.TypeIs](https://learn.microsoft.com/en-us/dotnet/api/system.linq.expressions.expression.typeis?view=net-7.0))

I can't explain with words much better, but if you are interested and have some time, please check out [this demo project](https://github.com/zackattackz/RazorPagesEFCoreFilterDemo) I put together, using my fork of [FilterExpressionCreator](https://github.com/zackattackz/FilterExpressionCreator/tree/SubclassFilters) (a similar library to Sieve)


Specifically, check out [this file in particular](https://github.com/zackattackz/RazorPagesEFCoreFilterDemo/blob/bdd806b757357b92680b83dedb00a9c2aad36c4d/RazorPagesEFCoreFilterDemo/Pages/Animals/List.cshtml.cs), which will generate the expression trees that I need. (check out the console output when accessing that page)

Also, here is [the issue](https://github.com/fschick/FilterExpressionCreator/issues/3) I opened in FilterExpressionCreator describing this issue and my feature in more depth.

I may just move forward with FilterExpressionCreator, but I'd like to contribute to Sieve if possible, unless this feature already exists.

**Comment by zackattackz** [comment]
`x => ((x Is Reptile) AndAlso Invoke(x => ((x.ScaleHardness > 10) AndAlso (((x Is Turtle) AndAlso Invoke(x => (x.ShellHardness > 10), Convert(x, Turtle))) OrElse ((x Is Crocodile) AndAlso Invoke(x => (x.BiteForce < 10), Convert(x, Crocodile))))), Convert(x, Reptile)))`

Is an example of the expression tree generated in /Animals/8 of that demo project.

See how it will check if  `x Is SubType` before then checking the `SubType` specific filters (combing them with `AndAlso`)

**Comment by hasanmanzak** [comment]
Yeah, I now understand what you're seeking for. Sieve does not have that kind of capability right now. I'm not sure if it needs to have 🙄 Currently there's no chaining of the subclasses, which this feature needs to have.

Erm, I really don't know how to follow-up on this. Since this is really is not an issue bu a feature request, I guess. Shall we go continue in a discussion thread?

**Comment by zackattackz** [comment]
Yeah it was more so a request. I don't really need the feature, was just suggesting. Figured someone in the future might need it. If it is, as you say, not necessary: What would you suggest someone who has an existing project with a class hierarchy do if they need this feature?

**Comment by hasanmanzak** [comment]
I didn't test it but my best guess is left joined queryables could do.
